src/helpers/getScValsFromSpec.ts

```diff
@@ -0,0 +1,66 @@
+import { contract, nativeToScVal, ScInt, xdr } from "@stellar/stellar-sdk";
+
+import { isSpecTypeNumber } from "@/helpers/isSpecTypeNumber";
+import { readObj } from "@/helpers/readObj";
+
+import { SorobanInvokeValue } from "@/types/types";
+
+// Used in the JsonSchemaForm component:
+// To convert the string value to ScVal
+// based on its function name and spec
+export const getScValsFromSpec = (
+  function_name: string,
+  spec: contract.Spec,
+  value: SorobanInvokeValue,
+) => {
+  const fn = spec.getFunc(function_name);
+
+  return fn.inputs().map((input: xdr.ScSpecFunctionInputV0) => {
+    const rawInputValue = readObj(value.args, input);
+    const { name: inputFieldType } = input.type().switch();
+
+    if (isSpecTypeNumber(input.type().switch())) {
+      const parsedIntValue = new ScInt(rawInputValue);
+      // without the following check, nativeToScVal makes it
+      // a U64 / I64 instead of a U32 / I32
+      if (inputFieldType === xdr.ScSpecType.scSpecTypeU32().name) {
+        return xdr.ScVal.scvU32(Number(rawInputValue));
+      }
+
+      if (inputFieldType === xdr.ScSpecType.scSpecTypeI32().name) {
+        return xdr.ScVal.scvI32(Number(rawInputValue));
+      }
+
+      if (inputFieldType === xdr.ScSpecType.scSpecTypeU64().name) {
+        return parsedIntValue.toU64();
+      }
+
+      if (inputFieldType === xdr.ScSpecType.scSpecTypeI64().name) {
+        return parsedIntValue.toI64();
+      }
+
+      if (inputFieldType === xdr.ScSpecType.scSpecTypeI128().name) {
+        return parsedIntValue.toI128();
+      }
+
+      if (inputFieldType === xdr.ScSpecType.scSpecTypeU128().name) {
+        return parsedIntValue.toU128();
+      }
+
+      return nativeToScVal(parsedIntValue, { type: inputFieldType });
```

Comment:
I think you can use `return parsedIntValue.toScVal()` here?

Reply:
`parseIntValue.toScVal` returns `u64` even for the required `u128` type. I think it's because the saved value type is in `string` type so it returns the smallest possible XDR integer type when I use `parsedIntValue.toScVal()`, which is `u64` in this case even though the contract requires `u128` in the case of swap func in contract `CA6PUJLBYKZKUEKLZJMKBZLEKP2OTHANDEOWSFF44FTSYLKQPIICCJBE`

Reply:
Ahhh yes that makes sense - you can pass an optional `{ type: "" }` but that'd be the same amount of code to compare the input field type.